warehouse/integrations/vulnerabilities/models.py

```diff
@@ -10,71 +10,65 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+import datetime
 
-from sqlalchemy import Column, ForeignKey, ForeignKeyConstraint, String, Table, orm
-from sqlalchemy.dialects.postgresql import ARRAY, TIMESTAMP
-from sqlalchemy.orm import mapped_column
+from sqlalchemy import ForeignKey, ForeignKeyConstraint, String, orm
+from sqlalchemy.dialects.postgresql import ARRAY
+from sqlalchemy.orm import Mapped, mapped_column
 
 from warehouse import db
+from warehouse.packaging.models import Release
 
-# TODO: convert to Declarative API
-release_vulnerabilities = Table(
-    "release_vulnerabilities",
-    db.metadata,
-    Column(
-        "release_id",
+
+class ReleaseVulnerability(db.ModelBase):
+    __tablename__ = "release_vulnerabilities"
+    __table_args__ = (
+        ForeignKeyConstraint(
+            ["vulnerability_source", "vulnerability_id"],
+            ["vulnerabilities.source", "vulnerabilities.id"],
+            onupdate="CASCADE",
+            ondelete="CASCADE",
+        ),
+    )
+
+    release_id: Mapped[str] = mapped_column(
         ForeignKey("releases.id", onupdate="CASCADE", ondelete="CASCADE"),
-        nullable=False,
         index=True,
-    ),
-    Column(
-        "vulnerability_source",
-        String,
-        nullable=False,
-    ),
-    Column(
-        "vulnerability_id",
-        String,
-        nullable=False,
-    ),
-    ForeignKeyConstraint(
-        ["vulnerability_source", "vulnerability_id"],
-        ["vulnerabilities.source", "vulnerabilities.id"],
-        onupdate="CASCADE",
-        ondelete="CASCADE",
-    ),
-)
+        primary_key=True,
+    )
+    vulnerability_source: Mapped[str]
+    vulnerability_id: Mapped[str]
 
 
 class VulnerabilityRecord(db.ModelBase):
     __tablename__ = "vulnerabilities"
 
-    source = mapped_column(String, primary_key=True)
-    id = mapped_column(String, primary_key=True)
+    source: Mapped[str] = mapped_column(String, primary_key=True)
+    id: Mapped[str] = mapped_column(primary_key=True)
 
     # The URL for the vulnerability report at the source
     # e.g. "https://osv.dev/vulnerability/PYSEC-2021-314"
-    link = mapped_column(String)
+    link: Mapped[str]
```

Comment:
Since the default column behavior is implicit `nullable=True`, we need to annotate the type that this could be null:
```suggestion
    link: Mapped[str | None]
```

Similar annotations should be made for the other implicit nullables in this class

Reply:
Wait, since this didn't originally have `nullable=True`, shouldn't this be `Mapped[str]` and not `Mapped[str | None]`? Doesn't "nullable" imply `Optional`?

Reply:
`Column.nullable` Defaults to `True`, which implies `Optional` aka `foo | None` (these are the same as of Python 3.10 / [PEP-604](https://peps.python.org/pep-0604/))

`Mapped[str]` == `nullable=False` (aka required), so to preserve the implicit `nullable=True`, we make denote it optional.

Reply:
Aha, got it!